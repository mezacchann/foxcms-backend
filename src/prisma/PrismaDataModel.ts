import { Injectable, Inject, Logger } from '@nestjs/common';
import * as fs from 'fs';
import * as util from 'util';
import outdent from 'outdent';

const appendFile = util.promisify(fs.appendFile);
const readFile = util.promisify(fs.readFile);
const writeFile = util.promisify(fs.writeFile);

@Injectable()
export class PrismaDataModel {
  private readonly logger = new Logger(PrismaDataModel.name, true);
  constructor(
    @Inject('ContentTypesDatamodel') private contentTypeDataModel: Buffer,
    @Inject('ContentTypeDataModelPath')
    private contentTypeDataModelPath: string,
  ) {}

  async addType(typeName: string) {
    if (this.typeExists(typeName))
      throw new Error(`Type ${typeName} already exists`);
    if (/\s/.test(typeName))
      throw new Error('Type name may not contain any whitespaces');
    try {
      await this.addTypeToDatamodel(typeName);
      this.reloadDatamodel();
      this.logger.log(`Added and deployed new content type ${typeName}`);
    } catch (err) {
      this.logger.error(err);
      throw new Error('Type cannot be added');
    }
  }

  async addField(
    contentTypeName: string,
    fieldName: string,
    fieldType: any,
    isRequired: boolean,
  ) {
    if (!this.typeExists(contentTypeName))
      throw new Error(
        `Cannot add Field. Type ${contentTypeName} doesn't exist`,
      );
    if (this.fieldExistsWithinType(contentTypeName, fieldName))
      throw new Error(
        `Field ${fieldName} exists already within type ${contentTypeName}`,
      );
    await this.addFieldToDatamodel(
      contentTypeName,
      fieldName,
      fieldType,
      isRequired,
    );
    this.reloadDatamodel();
  }

  private async addTypeToDatamodel(typeName: string) {
    const typeTemplate = outdent`\n
    type ${typeName} {
      id: ID! @unique
    }`;
    return appendFile(this.contentTypeDataModelPath, typeTemplate);
  }

  private async addFieldToDatamodel(
    contentTypeName: string,
    fieldName: string,
    fieldType: any,
    isRequired: boolean,
  ) {
    const fileContent = this.contentTypeDataModel.toString();
    const matchedType = fileContent.match(
      `type\\\s${contentTypeName}\\\s*\\{[^{}]*`,
    )[0];
    const idx = fileContent.indexOf(matchedType) + matchedType.length;
    const result =
      fileContent.slice(0, idx) +
      `  ${fieldName}: ${fieldType}${isRequired ? '!' : ''}\n` +
      fileContent.slice(idx);
    return writeFile(this.contentTypeDataModelPath, result);
  }

  async deleteType(contentTypeName: string) {
    if (!this.typeExists(contentTypeName))
      throw new Error(`Type ${contentTypeName} doesn't exists`);
    const fileContent = this.contentTypeDataModel.toString();
    const regex = new RegExp(
      `type\\\s*${contentTypeName}\\\s*\\{[^{}]*\\}\\\s`,
    );
    await writeFile(
      this.contentTypeDataModelPath,
      fileContent.replace(regex, ''),
    );
    this.reloadDatamodel();
  }

  async deleteContentTypeField(contentTypeName: string, fieldName: string) {
    if (!this.typeExists(contentTypeName))
      throw new Error(`Type ${contentTypeName} doesn't exists`);
    if (!this.fieldExistsWithinType(contentTypeName, fieldName))
      throw new Error(
        `Field ${fieldName} doesn't existP within type ${contentTypeName}`,
      );
    const fileContent = this.contentTypeDataModel.toString();
    const regex = new RegExp(`type.*${contentTypeName}\\\s*\\{[^{}]*\\}`);
    const matchedContent = fileContent.match(regex)[0];
    const typeWithRemovedField = matchedContent.replace(
      new RegExp(`[^\S\r\n]*${fieldName}.*\n`),
      '',
    );
    await writeFile(
      this.contentTypeDataModelPath,
      fileContent.replace(matchedContent, typeWithRemovedField),
    );
    this.reloadDatamodel();
  }

  private fieldExistsWithinType(typeName: string, fieldName: string) {
    const regex = new RegExp(`type.*${typeName}.*{([^}]+)}`);
    const result = this.contentTypeDataModel.toString().match(regex);
    return result[0].indexOf(fieldName) > -1;
  }

  private typeExists(typeName: string): boolean {
    const regex = new RegExp(`type.*${typeName}.*{`);
    const result = this.contentTypeDataModel.toString().match(regex);
    return result !== null;
  }

  private reloadDatamodel() {
    const fileContent = fs.readFileSync(this.contentTypeDataModelPath);
    this.contentTypeDataModel = fileContent;
  }

  getContentTypeDataModel(): Buffer {
    return Buffer.from(this.contentTypeDataModel);
  }
}
