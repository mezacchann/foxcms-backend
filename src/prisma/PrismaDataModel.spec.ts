import { dataModels } from './dataModels.provider';
import { PrismaDataModel } from './PrismaDataModel';
import { Test } from '@nestjs/testing';
import { readFileSync, writeFileSync } from 'fs';

describe('PrismaDataModel', () => {
  let prismaDataModel: PrismaDataModel;
  let deployFn: jest.Mock<T>;
  let updateRemoteModelFn: jest.Mock<T>;
  const contentTypeDataModelPath = './test/resources/contentTypes.graphql.txt';
  let initialDatamodelContent: Buffer;

  beforeAll(() => {
    initialDatamodelContent = readFileSync(contentTypeDataModelPath);
  });

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      providers: [
        PrismaDataModel,
        {
          provide: 'DynamicModelPath',
          useValue: contentTypeDataModelPath,
        },
        {
          provide: 'ContentTypesDatamodel',
          useValue: '',
        },
        {
          provide: 'DynamicModel',
          useValue: '',
        },
        {
          provide: 'PrismaEndpoint',
          useValue: 'dummyEndpoint',
        },
      ],
    }).compile();

    prismaDataModel = module.get<PrismaDataModel>(PrismaDataModel);
    updateRemoteModelFn = jest
      .spyOn(prismaDataModel, 'updateRemoteModel')
      .mockImplementation(() => '');
    deployFn = jest
      .spyOn(prismaDataModel, 'deploy')
      .mockImplementation(() => '');
  });

  describe('Add a new content type to the data model', () => {
    it('Should add a new type to the data model', () => {
      prismaDataModel.addType('photo');
      const expectedContent = readFileSync(
        './test/resources/contentTypes.addType.graphql.txt',
      );
      expect(prismaDataModel.getContent()).toBe(expectedContent.toString());
      expect(deployFn.mock.calls.length).toBe(1);
      expect(updateRemoteModelFn.mock.calls.length).toBe(1);
    });

    it('Should throw an error when adding the same content type twice', () => {
      prismaDataModel.addType('photo');
      expect(() => prismaDataModel.addType('pho to')).toThrow();
      expect(deployFn.mock.calls.length).toBe(1);
      expect(updateRemoteModelFn.mock.calls.length).toBe(1);
    });

    it('Should throw an error when the type name contains a whitespace', () => {
      expect(() => prismaDataModel.addType('pho to')).toThrow();
      expect(deployFn.mock.calls.length).toBe(0);
      expect(updateRemoteModelFn.mock.calls.length).toBe(0);
    });
  });

  describe('Add a new field to a content type', () => {
    it('Should add new fields to a content type', () => {
      prismaDataModel.addType('photo');
      prismaDataModel.addField('photo', 'width', 'Float', true);
      prismaDataModel.addField('photo', 'height', 'Float', true);
      prismaDataModel.addField('photo', 'size', 'Float', false);
      prismaDataModel.addField('photo', 'uri', 'String', true);
      const expectedContent = readFileSync(
        './test/resources/contentTypes.addField.graphql.txt',
      );
      expect(prismaDataModel.getContent()).toBe(expectedContent.toString());
      expect(deployFn.mock.calls.length).toBe(5);
      expect(updateRemoteModelFn.mock.calls.length).toBe(5);
    });

    it('Should throw an error when adding a field to a not existent content type', () => {
      expect(() =>
        prismaDataModel.addField('photo', 'uri', 'String', true),
      ).toThrowError();
      expect(deployFn.mock.calls.length).toBe(0);
      expect(updateRemoteModelFn.mock.calls.length).toBe(0);
    });

    it('Should throw an error when adding a field with the same name to the same content type', () => {
      prismaDataModel.addType('photo');
      prismaDataModel.addField('photo', 'width', 'Float', true);
      expect(() =>
        prismaDataModel.addField('photo', 'width', 'String', true),
      ).toThrow();
      expect(deployFn.mock.calls.length).toBe(2);
      expect(updateRemoteModelFn.mock.calls.length).toBe(2);
    });
  });

  describe('Delete a content type from the data model', () => {
    it('Delete the content type from the data model', () => {
      prismaDataModel.addType('photo');
      prismaDataModel.addType('photo2');
      prismaDataModel.addType('photo3');
      prismaDataModel.deleteType('photo2');
      const expectedContent = readFileSync(
        './test/resources/contentTypes.deleteType.graphql.txt',
      );
      expect(prismaDataModel.getContent()).toBe(expectedContent.toString());
      expect(deployFn.mock.calls.length).toBe(4);
      expect(updateRemoteModelFn.mock.calls.length).toBe(4);
    });

    it('Should throw an error when deleting an not existent content type', () => {
      prismaDataModel.addType('photo');
      expect(() => prismaDataModel.deleteType('photo2')).toThrow();
      expect(deployFn.mock.calls.length).toBe(1);
      expect(updateRemoteModelFn.mock.calls.length).toBe(1);
    });
  });

  describe('Delete a content type field from the data model', () => {
    it('Delete a content type field from the data model', () => {
      prismaDataModel.addType('photo');
      prismaDataModel.addField('photo', 'width', 'Float', true);
      prismaDataModel.addField('photo', 'height', 'Float', true);
      prismaDataModel.addField('photo', 'size', 'Float', false);
      prismaDataModel.addField('photo', 'uri', 'String', true);
      prismaDataModel.deleteContentTypeField('photo', 'height');
      prismaDataModel.deleteContentTypeField('photo', 'width');
      const expectedContent = readFileSync(
        './test/resources/contentTypes.deleteField.graphql.txt',
      );
      expect(prismaDataModel.getContent()).toBe(expectedContent.toString());
      expect(deployFn.mock.calls.length).toBe(7);
      expect(updateRemoteModelFn.mock.calls.length).toBe(7);
    });

    it('Throw an error when deleting an not existent content type field', () => {
      prismaDataModel.addType('photo');
      prismaDataModel.addField('photo', 'width', 'Float', true);
      prismaDataModel.addField('photo', 'height', 'Float', true);
      prismaDataModel.addField('photo', 'size', 'Float', false);
      prismaDataModel.addField('photo', 'uri', 'String', true);
      prismaDataModel.deleteContentTypeField('photo', 'height');

      expect(() =>
        prismaDataModel.deleteContentTypeField('photo', 'width2'),
      ).toThrow();
      expect(deployFn.mock.calls.length).toBe(6);
      expect(updateRemoteModelFn.mock.calls.length).toBe(6);
    });
  });

  afterEach(() => {
    writeFileSync(contentTypeDataModelPath, initialDatamodelContent);
  });
});
