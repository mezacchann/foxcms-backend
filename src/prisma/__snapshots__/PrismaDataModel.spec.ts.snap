// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrismaDataModel Add a new content type to the data model Should add a new type to the data model 1`] = `"type photo {id: ID! @unique}"`;

exports[`PrismaDataModel Add a new field to a content type Should add new fields to a content type 1`] = `"type photo {id: ID! @unique width: Float! height: Float! size: Float uri: String!}"`;

exports[`PrismaDataModel Delete a content type field from the data model Delete a content type field from the data model 1`] = `"type photo {id: ID! @unique width: Float! height: Float! size: Float uri: String!}"`;

exports[`PrismaDataModel Delete a content type from the data model Delete the content type from the data model 1`] = `"type photo {id: ID! @unique}type photo3 {id: ID! @unique}"`;

exports[`PrismaDataModel Update a content type Update the content type name 1`] = `"type photo {id: ID! @unique}type photo2_NEWNAME {id: ID! @unique}type photo3 {id: ID! @unique}"`;
